### ISSUE TITLE
Window Edge Snap

### ISSUE THREAD

**Issue by miketurn** [LMMS/lmms#3152]
Hello Everyone,

Basically, LMMS as with most DAWs use many little windows (mixer, sequencer, pattern, effects, instruments, etc.) that most users move around quite a bit, I feel a great feature to add would be the ability to have a way to "snap" the window edges to each other, as well as to the inner edge window of LMMS itself.
This would make it more convenient to move windows around more quickly and neatly.  For example many times I try to put say an effect window at the inner edge of of LMMS and when you go a tad out of bounds it calls up the scroll bars which is fine, but the ability to have the option to snap to that edge before it does so would be a nice addition.

Open to suggestions, but there are all kinds of minor settings that could be added to this "snap" feature as well....
1.) Ability to set the sensitivity of when the snap is activated, 
2.) Maybe the ability to have a setting that once items are snapped together there can be a way to have them stay linked together being able to move them as a group
3.) Obviously the ability to turn this "snap" feature off for those who may not want this.

As with many people that have ideas, I would hope to see this in a very near release, I feel it would be something that everyone would use regularly.
Does anyone else feel this would be a beneficial addition to LMMS?  Please share your opinions / ideas.

Thank You to the LMMS creators, and to anyone who reads this and has any information to share.

**Comment by BaraMGB** [comment]
Since qmdi subwindow doesn't have a stick feature it would not be an easy way to implement this.

**Comment by SirBothersome** [comment]
Not to mention that usefulness of "stay on top" function for windows...

**Comment by miketurn** [comment]
BaraMGB
You mean to say that the programming would not allow for something like this?

SirBothersome
Not sure what you mean here?
What does "stay on top" feature have to do with window snapping?

I don't know anything about programming but I would hope that there would be a way to some how achieve this.  Hard to believe that the code would be that limited that something like this couldn't be done?

Anyways, thank you guys for your responses, if anyone else has any opinions to share please do so.

**Comment by Spekular** [comment]
@miketurn almost anything is possible, it's more a question of how difficult it would be and whether or not it's worth the time and effort required. It's important to consider the opportunity cost of a feature, ie "what else could we get done instead?".

There's also a proposal for a single window UI that would make this a non-issue.

**Comment by tresf** [comment]
>  mean to say that the programming would not allow for something like this?

Here's a much smaller QMdiArea task for reference.  Note, this is only the CODE conversation. :) https://github.com/LMMS/lmms/pull/2516

**Comment by SirBothersome** [comment]
>SirBothersome
Not sure what you mean here?
What does "stay on top" feature have to do with window snapping?

@miketurn, not a thing sir, but the feature allows a very simple, fluid toggling experience that is great for on the fly mixing and parameter editing during playback. Since the windows don't disappear to the lowest level when you navigate away from them, the functionality is somewhat similar to your window snapping idea. Why "stay on top" isn't enabled by default. I dunno...

**Comment by tresf** [comment]
> >  Not to mention that usefulness of "stay on top" function for windows...

> What does "stay on top" feature have to do with window snapping


Albeit off-topic, they may be able to share a common context menu.

![screen shot 2016-12-16 at 10 10 05 am](https://cloud.githubusercontent.com/assets/6345473/21267337/d7bc9788-c377-11e6-8b42-7b5d924ee31f.png)

**Comment by miketurn** [comment]
@tresf 
No, not off-topic at all I agree that is a good idea for a place to put the "snap" settings
I honesting keep forgetting about that context menu.

Not really needed, but with my new craze of providing pictures, here is just a simple graphic where the RED rectangles represent the edges to snap to.
I would love to see all windows (effects / instruments) get this, but even if just the main components get this, it would still be pretty helpful.
![window snaps](https://cloud.githubusercontent.com/assets/24556082/21911892/ce9670aa-d8f0-11e6-8b09-cbb6ed79e463.JPG)
Hope This Helps
